File: leetcode-java/src/classify/tree/PathSumII.java
package classify.tree;

import tools.TreeNode;

import java.util.*;

public class PathSumII {
    private List<List<Integer>> res = new ArrayList<>();

    /**
     * 1ms      99.96%      O(n)
     * 39.4MB   71.80%      O(1)
     * DFS
     */
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        helper(new ArrayList<>(), root, targetSum);

        return res;
    }

    private void helper(List<Integer> level, TreeNode root, int targetSum) {
        if (root == null) {
            return;
        }

        if (root.left == null && root.right == null && targetSum == root.val) {
            level.add(root.val);
            res.add(new ArrayList<>(level));
            level.remove(level.size() - 1);
            return;
        }

        level.add(root.val);
        helper(level, root.left, targetSum - root.val);
        helper(level, root.right, targetSum - root.val);
        level.remove(level.size() - 1);
    }
}

File: leetcode-scala/src/dfs/medium/PathSumII.scala
package dfs.medium

import tools.TreeNode

/**
 * @Project leetcode
 * @Author Ravooo
 * @Since 2020/7/17 10:08
 * @Description PathSumII
 */
object PathSumII {

    var res = List[List[Int]]()

    /**
     * 600 ms   100.00%
     * 52.2 MB  100.00%
     * @param root
     * @param sum
     * @return
     */
    def pathSum(root: TreeNode, sum: Int): List[List[Int]] = {

        if (root == null) return List()
        if (root.left == null && root.right == null && root.value == sum) return List(List(root.value))

        val resLeft = pathSum(root.left, sum - root.value)
        val resRight = pathSum(root.right, sum - root.value)

        resLeft.map(_.prepended(root.value)) ++: resRight.map(_.prepended(root.value))
    }
}
